Bluetooth: host: Add find by UUID helper function for application.

joerchan opened the pull request:
Allow to pass attribute as NULL pointer when using notify or indicate by
UUID. This will use the entire handle value range to search for an
attribute with a matching UUID.
Document optional parameters, and clarify attr and uuid usage in the
variable declaration in the struct for clarification.

Expose a helper function to the application that searches the local
database for the given attribute from its UUID.
Provide arguments to limit the search that matches the service
declaration to make it easy to limit the search to a specific service.

Fixes: #22210

joerchan committed:
Bluetooth: host: Allow attribute as NULL with notify / indicate by UUID

Allow to pass attribute as NULL pointer when using notify or indicate by
UUID. This will use the entire handle value range to search for an
attribute with a matching UUID.
Document optional parameters, and clarify attr and uuid usage in the
variable declaration in the struct for clarification.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Add helper function to find an attribute from its uuid

Expose a helper function to the application that searches the local
database for the given attribute from its UUID.
Provide arguments to limit the search that matches the service
declaration to make it easy to limit the search to a specific service.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2254,34 +2254,49 @@ static bool gatt_find_by_uuid(struct notify_data *found,
+	data.attr = params->attr;
 	data.handle = bt_gatt_attr_get_handle(data.attr);
-	if (!data.handle) {
-		return -ENOENT;
-	}
 
 	/* Lookup UUID if it was given */
 	if (params->uuid) {
 		if (!gatt_find_by_uuid(&data, params->uuid)) {
 			return -ENOENT;
 		}
+	} else {
+		if (!data.handle) {
+			return -ENOENT;
+		}
```
Does data.handle need to be set before we call `gatt_find_by_uuid`? Otherwise I suggest moving the line
```
data.handle = bt_gatt_attr_get_handle(data.attr);
```
to the new placement of the check for `if (!data.handle) {` as well.

joerchan replied:
`gatt_find_by_uuid` uses `data.handle`.

jhedberg reviewed (approved):
Looks fine to me. Would it also be possible to get some sample code (extension of existing app or new app) to demonstrate the use and usefulness of the new API?

joerchan committed:
Bluetooth: samples: Use bt_gatt_find_by_uuid in peripheral sample

Demonstrate use of bt_gatt_find_by_uuid in peripheral sample. This
avoids the magical array index.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg Added use of this function to the peripheral sample. Also found the need to add `const` to the attr for this use case, due to the way services are declared when static.
